include/revng/Pipeline/Context.h

```diff
@@ -20,73 +20,68 @@
 
 namespace pipeline {
 
-/// A simple pair that holds a savable object a name meant to describe a object
-/// which lifetime must be than the context.
-class NamedGlobalReference {
-private:
-  llvm::StringRef Name;
-  SavableObjectBase *Obj;
-
-public:
-  NamedGlobalReference(llvm::StringRef Name, SavableObjectBase &Obj) :
-    Name(Name), Obj(&Obj) {}
-
-public:
-  llvm::StringRef name() const { return Name; }
-  SavableObjectBase &value() const { return *Obj; }
-};
-
 /// A class that contains every object that has a lifetime longer than a
 /// pipeline.
 ///
 /// This includes the kinds and the NamedGlobals that will be available in that
 /// context.
 class Context {
 private:
-  llvm::StringMap<SavableObjectBase *> Globals;
+  SavableGlobalsMap Globals;
+  llvm::StringMap<std::any> LesserGlobals;
   KindsRegistry TheKindRegistry;
 
 private:
   explicit Context(KindsRegistry Registry) :
     TheKindRegistry(std::move(Registry)) {}
-  Context(llvm::ArrayRef<NamedGlobalReference> Globals, KindsRegistry Registry);
 
 public:
   Context();
-  Context(llvm::ArrayRef<NamedGlobalReference> Globals);
 
 public:
-  static Context fromRegistry(llvm::ArrayRef<NamedGlobalReference> Globals,
-                              KindsRegistry Registry) {
-    return Context(Globals, std::move(Registry));
-  }
-
   static Context fromRegistry(KindsRegistry Registry) {
     return Context(std::move(Registry));
   }
 
-  template<typename... T>
-  static Context createFromGlobals(T &&...Args) {
-    using namespace std;
-    using namespace llvm;
-    SmallVector<NamedGlobalReference, 4> Globals({ forward<T>(Args)... });
-    return Context(Globals);
-  }
-
 public:
   const KindsRegistry &getKindsRegistry() const { return TheKindRegistry; }
 
   template<typename T>
   llvm::Expected<T *> getGlobal(llvm::StringRef Name) const {
-    auto Iter = Globals.find(Name);
-    if (Iter == Globals.end()) {
+    return Globals.get<T>(Name);
+  }
+
+  llvm::Error
+  serializeGlobal(llvm::StringRef GlobalName, llvm::raw_ostream &OS) const {
+    return Globals.serialize(GlobalName, OS);
+  }
+
+  llvm::Error deserializeGlobal(llvm::StringRef GlobalName,
+                                const llvm::MemoryBuffer &Buffer) {
+    return Globals.deserialize(GlobalName, Buffer);
+  }
+
+  template<typename T, typename... ArgsT>
+  void addGlobal(llvm::StringRef Name, ArgsT &&...Args) {
+    Globals.emplace<T>(Name, std::forward<T>(Args)...);
+  }
+
+  template<typename T>
+  void addLesserGlobal(llvm::StringRef Name, T &ToAdd) {
+    LesserGlobals.try_emplace(Name, std::make_any<T *>(&ToAdd));
+  }
```

Comment:
What about:
* `Global` -> `MutableGlobal`
* `LesserGlobal` -> `ImmutableGlobal` or `ConstGlobal`?

*Lesser* doesn't tell me much.

Reply:
yeah it was a tmp name i forgot to change

Reply:
i kept global for global and i simply renamed add and get lesser into add and get, since you are pretty much simply adding a pointer and retrieving it.